Add SSA subtitle import

jstebbins opened the pull request:
**Description of Change:**
Allow import of SSA/ASS files in a similar fashion to our current SRT import.

Only works with the CLI currently.

**Test on:**

- [x] Windows 10+  (via MinGW)
- [x] macOS 10.13+
- [x] Ubuntu Linux

jstebbins committed:
Add SSA subtitle import

jstebbins commented:
This PR changes some core code that affects SRT file processing.  So SRT import should be tested in the GUIs even though SSA import is only enabled in the CLI.

jstebbins committed:
decssasub: fix subtitle start/stop time offset

sr55 commented:
This will close #226

sr55 commented:
MinGW:
libhb/decssasub.o:decssasub.c:(.text+0xe4): undefined reference to `getline'
libhb/decssasub.o:decssasub.c:(.text+0x4ce): undefined reference to `getline'
collect2: error: ld returned 1 exit status
../libhb/module.rules:37: recipe for target 'libhb/hb.dll' failed
make: *** [libhb/hb.dll] Error 1

jstebbins committed:
decssasub: fix crash when filename is invalid

jstebbins committed:
LinGui: add SSA import

jstebbins committed:
Add hb_getline since mingw doesn't have getline function

sr55 commented:
Build now works thanks

jstebbins committed:
ports: move hb_getline to ports.c where it belongs

sr55 commented:
SRT Import still seems to behave as expected. 
SSA file shows this in the log:

[18:01:37]  * subtitle track 1, English [**SRT**] (track 0, id 0xff, Text) -> Passthrough, offset: 0, charset: UTF-8

jstebbins commented:
Not for me it dosn't
> [11:27:11]  * subtitle track 1, Unknown [SSA] (track 0, id 0xff, Text) -> Render/Burn-in

Are you using it right?  There are new CLI options --ssa-file etc...

sr55 commented:
Sorry, didn't realise you changed the JSON. Corrected and shows correctly in the log but segfaults on encode complete now. Looking into it. 

EDIT: SSA only.

jstebbins committed:
LinGui: fix position of subtitle radio buttons

sr55 commented:
"SubtitleList": [
      {
        "Burn": false,
        "Default": false,
        "Forced": false,
        "ID": 0,
        "Offset": 0,
        "Track": -1,
        "Import": {
          "Codeset": "UTF-8",
          "Format": "SSA",
          "Filename": "D:\\Development\\HandBrake\\Samples\\Real SSA.ssa",
          "Language": "eng"
        }
      }

Is this what you expect?


Getting 
[19:47:05]  * subtitle track 1, English [SSA] (track 0, id 0xff, Text) -> Passthrough, Forced Only
Failure to initialise thread 'SSA Subtitle Decoder'

Followed by a hard crash 

(gdb) bt
#0  0x00007fff8f0baed3 in ntdll!RtlIsNonEmptyDirectoryReparsePointAllowed () from C:\WINDOWS\SYSTEM32\ntdll.dll
1  0x00007fff8f0c379e in ntdll!RtlpNtMakeTemporaryKey () from C:\WINDOWS\SYSTEM32\ntdll.dll
2  0x00007fff8f0c3aaa in ntdll!RtlpNtMakeTemporaryKey () from C:\WINDOWS\SYSTEM32\ntdll.dll
3  0x00007fff8f05ebb1 in ntdll!RtlRaiseStatus () from C:\WINDOWS\SYSTEM32\ntdll.dll
4  0x00007fff8f06c3a9 in ntdll!memset () from C:\WINDOWS\SYSTEM32\ntdll.dll

jstebbins committed:
ports: fix memory corruption in hb_getline

jstebbins commented:
Crash should be fixed.

SubtitleList looks right

jstebbins commented:
FYI, Codeset isn't relevant for SSA so should be hidden in GUI when setting up an SSA import

sr55 commented:
Cheers, Confirmed Fixed.

MP4:  
![image](https://user-images.githubusercontent.com/628593/50741246-50640200-11f2-11e9-9a4c-3b36b0f80d68.png)

MKV:

![image](https://user-images.githubusercontent.com/628593/50741253-5ce85a80-11f2-11e9-925e-0f898fb99389.png)

There supposed to be a difference? Playing in VLC here.

jstebbins commented:
We have a bit of an omission in converting SSA styles to TX3G.  If the style is modified with a tag inline in the ssa subtitle, it gets converted to the appropriate tx3g style.  But SSA styles defined in the SSA header are not currently being processed.  So default colors are not getting set.

jstebbins committed:
ssa: improve SSA to TX3G conversion

We were only applying SSA inline override tags.  With this patch we now
parse SSA style descritions in the SSA header and apply them per event.

jstebbins committed:
ssautil: remove debug printf

jstebbins committed:
ssautil: fix strndup on mingw

sr55 commented:
LGTM.  Nice improvement in MP4 files too.
